<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>PanelPRO master function — PanelPRO • PanelPRO</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="PanelPRO master function — PanelPRO"><meta property="og:description" content="This is the main function of the PanelPRO package. It combines the
pre-processing steps (checking the user-supplied pedigree, imputing
missing ages, and setting up the database of model parameters) with the
calculation of the genotypic probability distributions and future risks of
cancer for the specified probands."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PanelPRO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/building-pedigrees.html">Building PanelPRO Pedigrees</a>
    </li>
    <li>
      <a href="../articles/rcpp-functions.html">Rcpp Functions</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>PanelPRO master function</h1>
    
    <div class="hidden name"><code>PanelPRO.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is the main function of the <code>PanelPRO</code> package. It combines the
pre-processing steps (checking the user-supplied pedigree, imputing
missing ages, and setting up the database of model parameters) with the
calculation of the genotypic probability distributions and future risks of
cancer for the specified probands.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">PanelPRO</span><span class="op">(</span></span>
<span>  <span class="va">pedigree</span>,</span>
<span>  model_spec <span class="op">=</span> <span class="st">"PanPRO22"</span>,</span>
<span>  proband <span class="op">=</span> <span class="va">pedigree</span><span class="op">$</span><span class="va">ID</span><span class="op">[</span><span class="va">pedigree</span><span class="op">$</span><span class="va">isProband</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cancers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  database <span class="op">=</span> <span class="va">PanelPRODatabase</span>,</span>
<span>  unknown.race <span class="op">=</span> <span class="va">UNKNOWN_RACE</span>,</span>
<span>  unknown.ancestry <span class="op">=</span> <span class="va">UNKNOWN_ANCESTRY</span>,</span>
<span>  impute.missing.ages <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  allow.intervention <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ignore.proband.germ <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max.mut <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove.miss.cancers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  iterations <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  max.iter.tries <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  debug <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  net <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  age.by <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  random.seed <span class="op">=</span> <span class="fl">42L</span>,</span>
<span>  plusBCRAT <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bcrat.vars <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rr.bcrat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rr.pop <span class="op">=</span> <span class="va">rr.ref</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>pedigree</dt>
<dd><p>A data frame of family history information with the
following columns. Unknown or missing values should be explicitly coded
as <code>NA</code>.</p><ul><li><p><code>ID</code>: A numeric value; ID for each individual. There should not be any
duplicated entries.</p></li>
<li><p><code>Sex</code>: A numeric value; <code>0</code> for female and <code>1</code> for male. Missing entries
are not currently supported.</p></li>
<li><p><code>MotherID</code>: A numeric value; unique ID for someone's mother.</p></li>
<li><p><code>FatherID</code>: A numeric value; unique ID for someone's father.</p></li>
<li><p><code>isProband</code>: A numeric value; <code>1</code> if someone is a proband, <code>0</code> otherwise.
This will be overridden by the <code>proband</code> argument in <code>PanelPRO</code>, if it is
specified. At least one proband should be specified by either the
<code>isProband</code> column or <code>proband</code>. Multiple probands are supported.</p></li>
<li><p><code>CurAge</code>: A numeric value; the age of censoring (current age if the person
is alive or age of death if the person is dead). Ages ranging from <code>1</code> to
<code>94</code> are allowed.</p></li>
<li><p><code>isAffX</code>: A numeric value; the affection status of cancer <code>X</code>, where <code>X</code>
is a <code>short</code> cancer code (see Details). Affection status should be encoded
as <code>1</code> if the individual was diagnosed, <code>0 </code>otherwise. Missing entries are
not currently supported.</p></li>
<li><p><code>AgeX</code>: A numeric value; the age of diagnosis for cancer <code>X</code>, where <code>X</code> is
a <code>short</code> cancer code (see Details). Ages ranging from <code>1</code> to <code>94</code> are
allowed. If the individual was not diagnosed for a given cancer, their
affection age should be encoded as <code>NA</code> and will be ignored otherwise.</p></li>
<li><p><code>isDead</code>: A numeric value; <code>1</code> if someone is dead, <code>0</code> otherwise. Missing
entries are assumed to be <code>0</code>.</p></li>
<li><p><code>race</code>: A character string; expected values are <code>"All_Races"</code>, <code>"AIAN"</code>
(American Indian and Alaska Native), <code>"Asian"</code>, <code>"Black"</code>, <code>"White"</code>,
<code>"Hispanic"</code>, <code>"WH"</code> (white Hispanic), and <code>"WNH"</code> (non-white Hispanic)
(see <code>PanelPRO:::RACE_TYPES</code>). Asian-Pacific Islanders should be encoded as
<code>"Asian"</code>. Race information will be used to select the cancer and death by
other causes penetrances used in the model. Missing entries are recoded as
the <code>unknown.race</code> argument, which defaults to
<code>PanelPRO:::UNKNOWN_RACE</code>.</p></li>
<li><p><code>Ancestry</code>: A character string; expected values are <code>"AJ"</code>(Ashkenazi Jewish),
<code>"nonAJ"</code>, and <code>"Italian"</code> (see <code>PanelPRO:::ANCESTRY_TYPES</code>). The ancestry
information is used to determine the allele frequencies of BRCA1 and BRCA2,
if they are included in the model. Missing entries are recoded as the
<code>unknown.ancestry</code> argument, which defaults to <code>PanelPRO:::UNKNOWN_ANCESTRY</code>.</p></li>
<li><p><code>Twins</code>: A numeric value; <code>0</code> for non-identical/single births, <code>1</code> for
the first set of identical twins/multiple births in the family, <code>2</code> for the
second set, etc. Missing entries are assumed to be <code>0</code>.</p></li>
<li><p><code>riskmod</code>: A character list; expected values are <code>"Mastectomy"</code>,
<code>"Hysterectomy"</code>, and <code>"Oophorectomy"</code>
(see <code>data.frame(PanelPRO:::RISKMODS)</code>). These preventative interventions
will be used to modify the cancer penetrances for breast, endometrial, and
ovarian cancer, respectively. Currently, <code>"Mastectomy"</code> refers to a
bilateral prophylactic mastectomy (occurring prior to first breast cancer).</p></li>
<li><p><code>interAge</code>: A numeric list; the age of intervention for each risk modifier
in <code>riskmod</code>. For example, <code>riskmod = list("Mastectomy", "Hysterectomy")</code>
and <code>interAge = {45, 60}</code> indicates that the individual had a mastectomy at
age 45 and a hysterectomy at age 60.</p></li>
<li><p>There can be optional columns for germline testing results (e.g. <code>BRCA1</code>,
<code>MLH1</code>) or tumor marker testing results. <code>ER</code>, <code>PR</code>, <code>CK14</code>, <code>CK5.6</code> and
<code>HER2</code> are tumor markers associated with breast cancer that will modify the
likelihood of phenotypes associated with <code>BRCA1</code> and <code>BRCA2</code>. <code>MSI</code> is a
tumor marker for colorectal cancer that will modify the likelihoods
associated with <code>MLH1</code>, <code>MSH2</code> <code>MSH6</code> and <code>PMS2</code>. For each of these optional
columns, positive results should be coded as <code>1</code>, negative results should be
coded as <code>0</code>, and unknown results should be coded as <code>NA</code>.</p></li>
</ul></dd>


<dt>model_spec</dt>
<dd><p>One of the following character strings indicating a
pre-specified model: <code>"BRCAPRO"</code>, <code>"BRCAPRO5"</code>, <code>"BRCAPRO6"</code>, <code>"MMRPRO"</code>,
<code>"PanPRO11"</code>, and <code>"PanPRO22"</code>. See <code>PanelPRO:::MODELPARAMS</code> for more
information. Either <code>model_spec</code> or <code>genes</code> and <code>cancers</code> should be
specified by the user. When both input types are supplied, the input from
<code>genes</code> and <code>cancers</code> will take precedence. The default is <code>"PanPRO22"</code>,
the largest pre-specified model currently supported by the package.</p></dd>


<dt>proband</dt>
<dd><p>A numeric value or vector of the unique IDs in <code>pedigree</code>
corresponding to the proband(s) for whom carrier probabilities and future
risk predictions should be estimated. The default is
<code>pedigree$ID[pedigree$isProband == 1]</code>, in which case probands will be
inferred from the <code>isProband</code> column in <code>pedigree</code>, if it exists and is
valid. Otherwise, the IDs in <code>proband</code> will override the contents of
<code>pedigree$isProband</code>. The user can also specify <code>proband = "All"</code>, in which
case all IDs in the <code>ped</code> will be treated as probands.</p></dd>


<dt>genes</dt>
<dd><p>A character vector of the genes of interest. The default is
<code>NULL</code>. Available options are listed in <code>PanelPRO:::GENE_TYPES</code>. Either
<code>model_spec</code> or <code>genes</code> and <code>cancers</code> should be specified by the user. When
both input types are supplied, the input from <code>genes</code> and <code>cancers</code> will
take precedence. If at least one cancer is specified in <code>cancers</code>, but no
genes are specified in <code>genes</code>, this will be treated as a valid model that
calculates future risk predictions based on the model parameters in
<code>database</code>, without incorporating the family history in <code>pedigree</code>, and does
not return any carrier probabilities.</p></dd>


<dt>cancers</dt>
<dd><p>A character vector of the cancers of interest. Available
options are listed in <code>PanelPRO:::CANCER_TYPES</code>. The default is <code>NULL</code>.
Either <code>model_spec</code> or <code>genes</code> and <code>cancers</code> should be specified by the
user. When both input types are supplied, the input from <code>genes</code> and
<code>cancers</code> will take precedence. If at least one gene is specified in
<code>genes</code>, but no cancers are specified in <code>cancers</code>, this will be treated as
a valid model that calculates carrier probability estimates based on the
model parameters in <code>database</code>, without incorporating the family history in
<code>pedigree</code>, and does not return any future risks.</p></dd>


<dt>database</dt>
<dd><p>A structured list providing the parameters of the model.
The default is <code><a href="PanelPRODatabase.html">PanelPRODatabase</a></code>. Users who wish to provide
their own model parameters should supply a list with the same structure as
<code>PanelPRODatabase</code>.</p></dd>


<dt>unknown.race</dt>
<dd><p>A character string indicating the default race to use
when <code>pedigree$race</code> is missing or unsupported. The default is
<code>PanelPRO:::UNKNOWN_RACE</code>.</p></dd>


<dt>unknown.ancestry</dt>
<dd><p>A character string indicating the default ancestry
to use when <code>pedigree$Ancestry</code> is missing or unsupported. The default is
<code>PanelPRO:::UNKNOWN_ANCESTRY</code>.</p></dd>


<dt>impute.missing.ages</dt>
<dd><p>A logical value indicating whether missing ages
should be multiply imputed. The default is <code>TRUE</code>. If set to <code>FALSE</code>, an
error will be raised if any individuals have missing current or cancer
affection ages.</p></dd>


<dt>allow.intervention</dt>
<dd><p>A logical value indicating if cancer penetrance
risk modifiers should be considered. The default is <code>TRUE</code>. Currently
supported interventions are preventative mastectomies, hysterectomies, and
oophorectomies (see <code>data.frame(PanelPRO:::RISKMODS)</code>).</p></dd>


<dt>ignore.proband.germ</dt>
<dd><p>A logical value indicating if the proband(s)'s
germline testing results (if provided in <code>ped</code>) should be considered. The
default is <code>FALSE</code>, in which case the proband(s)'s germline testing results
will not be ignored.</p></dd>


<dt>max.mut</dt>
<dd><p>The maximum number of simultaneous mutations allowed. The
default is <code>NULL</code>, in which case a maximum of <code>2</code> simultaneous mutations
will be used when <code>pedigree</code> has 30 or fewer members, and <code>1</code> otherwise.</p></dd>


<dt>remove.miss.cancers</dt>
<dd><p>A logical value indicating whether cancers in
<code>cancers</code> or implied by <code>model_spec</code> should be removed from the model
specification when they are missing in <code>pedigree</code>. The default is <code>TRUE</code>.</p></dd>


<dt>iterations</dt>
<dd><p>A numeric value indicating the number of iterations that
should be run when multiply imputing missing ages. The default is <code>20</code>.</p></dd>


<dt>max.iter.tries</dt>
<dd><p>A numeric value indicating the maximum number of
iterations that should be attempted when multiply imputing missing ages.
(Invalid iterations typically occur when an individual is missing both their
current and cancer ages. Current ages for living individuals are imputed
first; if a valid cancer affection cannot be found, subject to the upper
bound set by the imputed current age, this imputation run will be
discarded.) The default is <code>NULL</code>, in which case the maximum number of
tries will be set to five times <code>iterations</code>.</p></dd>


<dt>parallel</dt>
<dd><p>A logical value indicating whether the peeling-paring
algorithm should be run in parallel when missing ages are imputed multiple
times. The default is <code>TRUE</code>.</p></dd>


<dt>debug</dt>
<dd><p>A logical value indicating whether to debug using browsers and
sequential imputation. The default is <code>FALSE</code>. Setting <code>debug = TRUE</code> is
only useful for package developers.</p></dd>


<dt>net</dt>
<dd><p>A logical value indicating whether to return net or crude future
risk estimates. The default is <code>FALSE</code>, in which case crude future risk
estimates will be returned. Net risk is the probability of developing a
given cancer by the specified age, in a world where there is no death from
other causes. Crude risk is the probability of developing a given cancer by
the specified age and not having died from other causes beforehand.</p></dd>


<dt>age.by</dt>
<dd><p>The fixed age interval for the future risk calculation. The
default is <code>5</code>, in which case if the proband's current age is 30, risk
predictions will be calculated at ages 35, 40, 45, etc.</p></dd>


<dt>random.seed</dt>
<dd><p>The random seed (a numeric value) to set for imputing
missing ages. The default is <code>42L</code>.</p></dd>


<dt>plusBCRAT</dt>
<dd><p>A logical value indicating whether to use the BRCAPRO+BCRAT
model. The default is <code>FALSE</code>. If <code>plusBCRAT = TRUE</code>, the user should
specify one of <code>bcrat.vars</code> or <code>rr.bcrat</code>. The BCRAPRO+BCRAT model is only
run if the user also sets <code>net = FALSE</code> and breast cancer is included in the
model specification.</p></dd>


<dt>bcrat.vars</dt>
<dd><p>A data frame of BCRAT covariates, where each row
corresponds to a proband in <code>pedigree</code>. The expected columns are:</p><ul><li><p><code>ID</code>: Numeric ID for each proband. Should not contain duplicated
entries.</p></li>
<li><p><code>Age</code>: Initial age (should be &gt;=20 and &lt;90).</p></li>
<li><p><code>NBiops</code>: Number of benign breast biopsies (<code>99</code> if unknown).</p></li>
<li><p><code>Hyp</code>: A logical value indicating whether the individual has had a breast
biopsy with atypical hyperplasia (<code>0</code> if no, <code>1</code> if yes, <code>99</code> if unknown).</p></li>
<li><p><code>AgeMen</code>: Age at menarche (<code>99</code> if unknown)</p></li>
<li><p><code>AgeFLB</code>: Age at first live birth (<code>98</code> if nulliparous, <code>99</code> if unknown).</p></li>
<li><p><code>NumRel</code>: The number of female first-degree relatives with breast cancer
(<code>99</code> if unknown).</p></li>
<li><p><code>Race</code>: A character string indicating race (<code>"White"</code>, <code>"Black"</code>,
<code>"Hispanic"</code>, <code>"FHispanic"</code>, <code>"Asian"</code>, or <code>"Other"</code>; <code>"Hispanic"</code>
corresponds to U.S.-born Hispanic while <code>"FHispanic"</code> corresponds to
foreign-born Hispanic).</p></li>
</ul><p>The default is <code>NULL</code>. <code>bcrat.vars</code> is only used if <code>plusBCRAT = TRUE</code>,
<code>net = FALSE</code>, breast cancer is included in the model specification, and
the user does not input <code>rr.bcrat</code>. The user only needs to input one of
<code>rr.bcrat</code> and <code>bcrat.vars</code> to run BCRAPRO+BCRAT; if both are specified,
<code>rr.bcrat</code> will be used for the model.</p></dd>


<dt>rr.bcrat</dt>
<dd><p>A data frame with 3 columns: the IDs of the probands (<code>ID</code>),
the probands' BCRAT relative risks for age &lt; 50 (<code>rr1</code>), and the probands'
BCRAT relative risks for age &gt;= 50 (<code>rr2</code>). When <code>net = FALSE</code> and breast
cancer is included in the model, these values will modify the future risk
calculation. The default is <code>NULL</code>. <code>rr.bcrat</code> is only used if
<code>plusBCRAT = TRUE</code>, <code>net = FALSE</code>, and breast cancer is included in the
model specification. The user only needs to input one of <code>rr.bcrat</code> and
<code>bcrat.vars</code> to run BCRAPRO+BCRAT; if both are specified, <code>rr.bcrat</code> will be
used for the model.</p></dd>


<dt>rr.pop</dt>
<dd><p>A data frame of race-specific for estimates of 1 /
(1 - population attributable risk). The default is <code>rr.ref</code>,
which was estimated from NHIS 2015.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with two components:</p><ul><li><p><code>posterior.prob</code>: A list where each element corresponds to a proband in
<code>proband</code> and consists of a data frame of posterior carrier probabilities.
There are three columns: <code>estimate</code> (the estimate if no ages were imputed,
or the estimate averaged over all age imputations), <code>lower</code> (<code>NA</code> if no ages
were imputed, or the minimum result for all age imputations), and <code>upper</code>
(<code>NA</code> if no ages were imputed, or the maximum result for all age
imputations). The rows correspond to the different genotypes. If no genes
were in the model specification, each proband's data frame of carrier
probabilities is replaced by the message
<code>"No carrier probabilities were requested by the model specification."</code>
If ages were imputed, <code>ImputeRange</code> messages giving the lower and upper
bounds for the probability of carrying any pathogenic variant will be
printed for each proband.
A <code>ZeroProb</code> warning is raised if one or more of the estimated
probabilities is <code>0</code> (or <code>1</code>), germline testing results were incorporated,
and default sensitivities/specifities of <code>1</code> were used.</p></li>
<li><p><code>future.risk</code>: A list where each element corresponds to a proband in
<code>proband</code> and consists of a list of data frames of future risk estimates at
fixed intervals for each cancer in the model. Each data frame has three
columns: <code>estimate</code> (the estimate if no ages were imputed, or the estimate
averaged over all age imputations), <code>lower</code> (<code>NA</code> if no ages were imputed,
or the minimum result for all age imputations), and <code>upper</code> (<code>NA</code> if no
ages were imputed, or the maximum result for all age imputations). Each row
corresponds to the risk estimate for at a different age, e.g. 35, 40, 45,
etc. when <code>age.by = 5</code> for a proband whose current age is 30. If no cancers
were in the model specification, each proband's list of future risk
estimates is replaced by the message
<code>"No future risk estimates were requested by the model specification."</code></p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>The <code>PanelPRO</code> master function combines pre-processing of model parameters
and the pedigree with calculation of the posterior carrier probabilities and
future risk estimates. First, a model-specific database is built by
subsetting <code>database</code> for the genes and cancers implied by <code>model_spec</code> or
<code>genes</code> and <code>cancers</code> (see <code><a href="buildDatabase.html">buildDatabase</a></code>). Then, <code>pedigree</code> is
checked for having a valid, consistent structure, and missing ages are
imputed (see <code><a href="checkFam.html">checkFam</a></code>). If <code>pedigree</code> fails a check,
corrections will be made when possible, and an informative error message
will be issued if not. Finally, the carrier probabilities and future risk
predictions for the user-specified probands are estimated and returned
(see <code><a href="PanelPROCalc.html">PanelPROCalc</a></code>).</p>
<p>For convenience, the <code>PanelPRO</code> package exports wrapper functions for all
of the pre-specified models: <code><a href="BRCAPRO.html">BRCAPRO</a></code>, <code><a href="BRCAPRO5.html">BRCAPRO5</a></code>,
<code><a href="BRCAPRO6.html">BRCAPRO6</a></code>, <code><a href="PanelPRO11.html">PanelPRO11</a></code>, and
<code><a href="PanelPRO22.html">PanelPRO22</a></code>.</p>
<p>To represent a cancer in <code>pedigree</code>, we need to use <code>short</code> cancer codes.
See <code>data.frame(PanelPRO:::CANCER_NAME_MAP)</code> for available options. For
example, <code>BC</code> is the <code>short</code> name that maps to <code>Breast</code>, so breast cancer
affection status and age of diagnosis should be recorded in columns
named <code>isAffBC</code> and <code>AgeBC</code>.</p>
<p>When the corresponding cancer penetrance information is available in
<code>PanelPRODatabase</code>, <code>PanelPRO</code> calculates and outputs carrier probabilities
for both homozygous and heterozygous mutations of the genes included in model.
MUTYH is the only gene for which we currently have both homozygous and
heterozygous penetrances; homozygous mutations for all other genes are
considered non-viable (see <code>PanelPRO:::ALL_GENE_VARIANT_TYPES</code>).</p>
<p>Sets of identical twins/multiple births only contribute to the likelihood
once; see <code><a href="dot-twinsLikMod.html">.twinsLikMod</a></code> for more information. Carrier
probabilities returned for individuals in the same set will be identical.</p>
<p>The R package <code>cbcrisk</code>, which is on GitHub but is not on CRAN, will be
automatically installed when the function runs, if is is not installed already.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="buildDatabase.html">buildDatabase</a></code>, <code><a href="checkFam.html">checkFam</a></code>,
<code><a href="PanelPROCalc.html">PanelPROCalc</a></code>, <code><a href="BRCAPRO.html">BRCAPRO</a></code>, <code><a href="BRCAPRO5.html">BRCAPRO5</a></code>,
<code><a href="BRCAPRO6.html">BRCAPRO6</a></code>, <code><a href="PanelPRO11.html">PanelPRO11</a></code>, <code><a href="PanelPRO22.html">PanelPRO22</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">PanelPRO</span><span class="op">(</span><span class="va">test_fam_1</span>, model_spec <span class="op">=</span> <span class="st">"BRCAPRO"</span>, proband <span class="op">=</span> <span class="fl">5</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Your model has 2 cancers - Breast, Ovarian and 2 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting max.mut to 2.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `proband` argument does not match with "isProband" column in the pedigree, using `proband`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 6,29 has tumor marker testing results but is unaffected for relevant cancer, so testing will be ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Germline testing results for BRCA1 are assumed to be for default variant BRCA1_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Germline testing results for BRCA2 are assumed to be for default variant BRCA2_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 3 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 10 's race has been changed to All_Races to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 9 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 16,17 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 29,30 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 33,34 's race has been changed to All_Races to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 33,34 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 5's lower and upper bounds for the probability of carrying any pathogenic variant are (1, 1).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Some of the estimated probabilities are 0 or 1, most likely due to the incorporation of reported germline testing results (by default, we assume that the sensitivity and specificity for all germline tests is 1). However, this does not guarantee that the person is a noncarrier or mutation carrier; a more accurate estimate may be achieved with user-specified sensitivities and specificities.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $posterior.prob</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $posterior.prob$`5`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                   genes   estimate      lower       upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                            noncarrier 0.00000000 0.00000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                    BRCA1_hetero_anyPV 0.99867676 0.99857092 0.998725730</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                    BRCA2_hetero_anyPV 0.00000000 0.00000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 BRCA1_hetero_anyPV.BRCA2_hetero_anyPV 0.00132324 0.00127427 0.001429079</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`$Breast</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ByAge     estimate        lower        upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    70 0.0002896075 0.0002880944 0.0002928779</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    75 0.0006028101 0.0005997432 0.0006094383</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    80 0.0009191193 0.0009145951 0.0009288971</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    85 0.0011888034 0.0011830966 0.0012011371</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    90 0.0013368604 0.0013304304 0.0013507573</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`$Ovarian</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ByAge estimate lower upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    70        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    75        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    80        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    85        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    90        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">PanelPRO</span><span class="op">(</span><span class="va">test_fam_1</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"BRCA2"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>         cancers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Breast"</span>, <span class="st">"Ovarian"</span><span class="op">)</span>, proband <span class="op">=</span> <span class="fl">5</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Your model has 2 cancers - Breast, Ovarian and 2 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting max.mut to 2.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `proband` argument does not match with "isProband" column in the pedigree, using `proband`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 6,29 has tumor marker testing results but is unaffected for relevant cancer, so testing will be ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Germline testing results for BRCA1 are assumed to be for default variant BRCA1_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Germline testing results for BRCA2 are assumed to be for default variant BRCA2_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 3 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 10 's race has been changed to All_Races to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 9 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 16,17 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 29,30 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 33,34 's race has been changed to All_Races to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 33,34 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 5's lower and upper bounds for the probability of carrying any pathogenic variant are (1, 1).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Some of the estimated probabilities are 0 or 1, most likely due to the incorporation of reported germline testing results (by default, we assume that the sensitivity and specificity for all germline tests is 1). However, this does not guarantee that the person is a noncarrier or mutation carrier; a more accurate estimate may be achieved with user-specified sensitivities and specificities.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $posterior.prob</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $posterior.prob$`5`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                   genes   estimate      lower       upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                            noncarrier 0.00000000 0.00000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                    BRCA1_hetero_anyPV 0.99867676 0.99857092 0.998725730</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                    BRCA2_hetero_anyPV 0.00000000 0.00000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 BRCA1_hetero_anyPV.BRCA2_hetero_anyPV 0.00132324 0.00127427 0.001429079</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`$Breast</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ByAge     estimate        lower        upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    70 0.0002896075 0.0002880944 0.0002928779</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    75 0.0006028101 0.0005997432 0.0006094383</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    80 0.0009191193 0.0009145951 0.0009288971</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    85 0.0011888034 0.0011830966 0.0012011371</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    90 0.0013368604 0.0013304304 0.0013507573</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $future.risk$`5`$Ovarian</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ByAge estimate lower upper</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    70        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    75        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    80        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    85        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    90        0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="PanelPRO22.html">PanelPRO22</a></span><span class="op">(</span><span class="va">test_fam_2</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span>, max.mut <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Brain, Breast, Colorectal, Gastric, Kidney, Leukemia, Melanoma, Ovarian, Osteosarcoma, Prostate, Soft Tissue Sarcoma, Thyroid, Urinary Bladder, Hepatobiliary, Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Your model has 3 cancers - Endometrial, Pancreas, Small Intestine and 22 genes - ATM_hetero_anyPV, BARD1_hetero_anyPV, BRCA1_hetero_anyPV, BRCA2_hetero_anyPV, BRIP1_hetero_anyPV, CDH1_hetero_anyPV, CDK4_hetero_anyPV, CDKN2A[P16]_hetero_anyPV, CHEK2_hetero_1100delC, EPCAM_hetero_anyPV, MLH1_hetero_anyPV, MSH2_hetero_anyPV, MSH6_hetero_anyPV, MUTYH_hetero_anyPV, MUTYH_homo_anyPV, NBN_hetero_657del5, PALB2_hetero_anyPV, PMS2_hetero_anyPV, PTEN_hetero_anyPV, RAD51C_hetero_anyPV, RAD51D_hetero_anyPV, STK11_hetero_anyPV, TP53_hetero_anyPV.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 12's lower and upper bounds for the probability of carrying any pathogenic variant are (0.17495, 0.25534).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 19's lower and upper bounds for the probability of carrying any pathogenic variant are (0.10484, 0.14514).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ID 21's lower and upper bounds for the probability of carrying any pathogenic variant are (0.04897, 0.05066).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span>
<span class="r-in"><span><span class="co"># Visualise the risk profile of the proband(s)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="visRisk.html">visRisk</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by BayesMendel Lab, Theo Huang, Gavin Lee, Jane Liang, Tom Madsen, Qing Zhang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

