<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PanelPRO: multi-syndrome and multi-gene modeling • PanelPRO</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="PanelPRO: multi-syndrome and multi-gene modeling">
<meta property="og:description" content="An R package for multi-gene multi-syndrome modeling from family history.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">PanelPRO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/building-pedigrees.html">Building PanelPRO Pedigrees</a>
    </li>
    <li>
      <a href="articles/rcpp-functions.html">Rcpp Functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="panelpro-r-package">PanelPRO R package<a class="anchor" aria-label="anchor" href="#panelpro-r-package"></a>
</h1></div>
<p>The PanelPRO R package extends the BayesMendel R package to multi-syndrome, multi-gene Mendelian risk prediction modeling.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="install-the-cbcrisk-r-package">Install the cbcrisk R package<a class="anchor" aria-label="anchor" href="#install-the-cbcrisk-r-package"></a>
</h3>
<p>The package cbcrisk version 2.0 is a dependency of PanelPRO v1.0.0 and higher. Prior to using PanelPRO, you can install it manually using the instructions below otherwise, it will install automatically from GitHub the first time you run the <code><a href="reference/PanelPRO.html">PanelPRO()</a></code> function or another model function (ie <code><a href="reference/BRCAPRO.html">BRCAPRO()</a></code>).</p>
<div class="section level4">
<h4 id="installation-via-a-targz-or-zip-file">Installation via a tar.gz or .zip file:<a class="anchor" aria-label="anchor" href="#installation-via-a-targz-or-zip-file"></a>
</h4>
<p>The tar.gz and .zip files for cbcrisk v2.0 can be found <a href="https://personal.utdallas.edu/~sxb125731/" class="external-link">here</a>.</p>
<p>On Linux, enter the following into your terminal:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>R CMD INSTALL CBCRisk_2.<span class="fl">0.</span>tar.gz</span></code></pre></div>
<p>On Windows, enter following into our R console:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"[PATH]/CBCRisk_2.0.zip"</span>, <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">"source"</span>)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="installation-via-github">Installation via GitHub<a class="anchor" aria-label="anchor" href="#installation-via-github"></a>
</h4>
<p>On GitHub, cbcrisk v2.0 can be found <a href="https://github.com/ihsajal/CBCRisk" class="external-link">here</a>. Note that the commit SHA in the code below is confirmed to be compatible with PanelPRO as of Feb. 23rd, 2023.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"ihsajal/CBCRisk@44f7b5ee801a0b4e09977e3f5645e17a38cc7598"</span>)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="install-panelpro-from-source">Install PanelPRO from source<a class="anchor" aria-label="anchor" href="#install-panelpro-from-source"></a>
</h3>
<p>If you are installing PanelPRO from source and running Linux, enter the following into the terminal:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>R CMD INSTALL PanelPRO_X.X.X.tar.gz</span></code></pre></div>
<p>For Windows, enter the following into your R console:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"[your PanelPRO file path]/PanelPRO_X.X.X.zip"</span>, <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">"source"</span>)</span></code></pre></div>
<p>where <code>X.X.X</code> corresponds to the version number you have downloaded.</p>
</div>
</div>
<div class="section level2">
<h2 id="quick-start-guide">Quick-start guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<p>This following is a quick-start guide for basic usage of the <code>PanelPRO</code> package. For greater detail on model options, please refer to the other vignettes and documentation, e.g <code><a href="articles/building-pedigrees.html">vignette("building-pedigrees", package = "PanelPRO")</a></code> or <code><a href="reference/PanelPRO.html">help(PanelPRO)</a></code>.</p>
<p>The primary function in the package is the eponymous <code>PanelPRO</code>, which outputs the posterior carrier probabilities and future risk of cancer of one or more probands, based on the user-supplied <code>pedigree</code> and model specification (either <code>model_spec</code> or explicit specification of <code>genes</code> and <code>cancers</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PanelPRO</span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to the PanelPRO package, by the BayesMendel Lab!</span></span>
<span><span class="co">#&gt; PanelPRO version: 1.0.0</span></span>
<span><span class="co">#&gt;    ___                _____  ___  ____  </span></span>
<span><span class="co">#&gt;   / _ \___ ____  ___ / / _ \/ _ \/ __ \ </span></span>
<span><span class="co">#&gt;  / ___/ _ `/ _ \/ -_) / ___/ , _/ /_/ / </span></span>
<span><span class="co">#&gt; /_/   \_,_/_//_/\__/_/_/  /_/|_|\____/</span></span></code></pre></div>
<div class="section level3">
<h3 id="pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#pedigree"></a>
</h3>
<p>The user must specify the <code>pedigree</code> argument as a data frame with the following columns:</p>
<ul>
<li>
<code>ID</code>: A numeric value; ID for each individual. There should not be any duplicated entries.</li>
<li>
<code>Sex</code>: A numeric value; <code>0</code> for female and <code>1</code> for male. Missing entries are not currently supported.</li>
<li>
<code>MotherID</code>: A numeric value; unique ID for someone’s mother.</li>
<li>
<code>FatherID</code>: A numeric value; unique ID for someone’s father.</li>
<li>
<code>isProband</code>: A numeric value; <code>1</code> if someone is a proband, <code>0</code> otherwise. This will be overridden by the <code>proband</code> argument in <code>PanelPRO</code>, if it is specified. At least one proband should be specified by either the <code>isProband</code> column or <code>proband</code>. Multiple probands are supported.</li>
<li>
<code>CurAge</code>: A numeric value; the age of censoring (current age if the person is alive or age of death if the person is dead). Ages ranging from <code>1</code> to <code>94</code> are allowed.</li>
<li>
<code>isAffX</code>: A numeric value; the affection status of cancer <code>X</code>, where <code>X</code> is a <code>short</code> cancer code (see below). Affection status should be encoded as <code>1</code> if the individual was diagnosed, <code>0</code>otherwise. Missing entries are not currently supported.</li>
<li>
<code>AgeX</code>: A numeric value; the age of diagnosis for cancer <code>X</code>, where <code>X</code> is a <code>short</code> cancer code (see below). Ages ranging from <code>1</code> to <code>94</code> are allowed. If the individual was not diagnosed for a given cancer, their affection age should be encoded as <code>NA</code> and will be ignored otherwise.</li>
<li>
<code>isDead</code>: A numeric value; <code>1</code> if someone is dead, <code>0</code> otherwise. Missing entries are assumed to be <code>0</code>.</li>
<li>
<code>race</code>: A character string; expected values are <code>"All_Races"</code>, <code>"AIAN"</code> (American Indian and Alaska Native), <code>"Asian"</code>, <code>"Black"</code>, <code>"White"</code>, <code>"Hispanic"</code>, <code>"WH"</code> (white Hispanic), and <code>"WNH"</code> (non-white Hispanic) (see <code>PanelPRO:::RACE_TYPES</code>). Asian-Pacific Islanders should be encoded as <code>"Asian"</code>. Race information will be used to select the cancer and death by other causes penetrances used in the model. Missing entries are recoded as the <code>unknown.race</code> argument, which defaults to <code>PanelPRO:::UNKNOWN_RACE</code>.</li>
<li>
<code>Ancestry</code>: A character string; expected values are <code>"AJ"</code>, <code>"nonAJ"</code>, and <code>"Italian"</code> (see <code>PanelPRO:::ANCESTRY_TYPES</code>). Ancestry information will be used to select the allele frequencies used in the model. Missing entries are recoded as the <code>unknown.ancestry</code> argument, which defaults to <code>PanelPRO:::UNKNOWN_ANCESTRY</code>.</li>
<li>
<code>Twins</code>: A numeric value; <code>0</code> for non-identical/single births, <code>1</code> for the first set of identical twins/multiple births in the family, <code>2</code> for the second set, etc. Missing entries are assumed to be <code>0</code>.</li>
<li>Prophylactic surgical history columns: as of PanelPRO v1.1.0, there are two ways to encode prophylactic surgical history for mastectomies, hysterectomies, and oophorectomies which are explained below. These preventative interventions will be used to modify the cancer penetrances for breast, endometrial, and ovarian cancer, respectively. Note that in order to be considered prophylactic, mastectomies and oophorectomies must have been bilateral.
<ul>
<li>The first method has been utilized since the initial release of PanelPRO and uses two list type columns:
<ul>
<li>
<code>riskmod</code>: A character list; expected values are <code>"Mastectomy"</code>, <code>"Hysterectomy"</code>, and <code>"Oophorectomy"</code> (see <code>data.frame(PanelPRO:::RISKMODS)</code>).</li>
<li>
<code>interAge</code>: A numeric list; the age of intervention for each risk modifier in <code>riskmod</code>. For example, <code>riskmod = list("Mastectomy", "Hysterectomy")</code> and <code>interAge = {45, 60}</code> indicates that the individual had a mastectomy at age 45 and a hysterectomy at age 60. Unknown surgery ages should be <code>NA</code>.</li>
</ul>
</li>
<li>The second method, which is only compatible with PanelPRO v1.1.0 or higher, uses six different numeric columns to replace the <code>riskmod</code> and <code>interAge</code> list type columns. This newer method makes sharing and storing pedigrees as .csv files easier because list type columns are not compatible with .csv files. The naming conventions for the six new columns are <code>riskmodX</code> (3 columns) and <code>interAgeX</code> (3 columns), where “X” is one of the 3 abbreviated surgery names <code>"Mast"</code>, <code>"Ooph"</code>, or <code>"Hyst"</code>:
<ul>
<li>
<code>riskmodX</code> A binary value; <code>1</code> if the surgery occurred and <code>0</code> otherwise.</li>
<li>
<code>interAgeX</code> A numeric value; the age of the corresponding intervention for <code>riskmodX</code>. For example, if <code>riskmodMast = 1</code> and <code>interAgeMast = 45</code> the individual had a bilateral prophylactic mastectomy at age 45. Unknown surgery ages should be coded as <code>NA</code>.</li>
</ul>
</li>
</ul>
</li>
<li>There can be optional columns for germline testing results (e.g. <code>BRCA1</code>, <code>MLH1</code>) or tumor marker testing results. <code>ER</code>, <code>PR</code>, <code>CK14</code>, <code>CK5.6</code> and <code>HER2</code> are tumor markers associated with breast cancer that will modify the likelihood of phenotypes associated with <code>BRCA1</code> and <code>BRCA2</code>. <code>MSI</code> is a tumor marker for colorectal cancer that will modify the likelihoods associated with <code>MLH1</code>, <code>MSH2</code> <code>MSH6</code> and <code>PMS2</code>. For each of these optional columns, positive results should be coded as <code>1</code>, negative results should be coded as <code>0</code>, and unknown results should be coded as <code>NA</code>.</li>
<li>There are optional columns related to breast cancer which provide more accurate estimates for contralateral breast cancer risk. These columns are:
<ul>
<li>
<code>FirstBCType</code>: Breast cancer type of the first primary breast cancer. The only options are <code>Invasive</code> for pure invasive and <code>Invasive_DCIS</code> for mixed invasive and DCIS. There is no option for pure DCIS (see “Additional Notes” section). Use <code>NA</code> if no first primary breast cancer or unknown type.</li>
<li>
<code>AntiEstrogen</code>: Anti-estrogen therapy used to treat first primary breast cancer. The options are <code>0</code> for no, <code>1</code> for yes, and <code>NA</code> for unknown or for those never diagnosed with breast cancer.</li>
<li>
<code>HRPreneoplasia</code>: History of high risk preneoplasia (ie atypical hyperplasia or LCIS). The options are <code>0</code> for no, <code>1</code> for yes, and <code>NA</code> for unknown or for those never diagnosed with breast cancer.</li>
<li>
<code>BreastDensity</code>: BI-RADS breast density descriptor. The options are the letters <code>a</code> through <code>d</code>. <code>a</code> = almost entirely fatty, <code>b</code> = scattered areas of fibroglandular density, <code>c</code> = heterogeneously dense, and <code>d</code> = extremely dense. If the breast density is unknown, use <code>NA</code>.</li>
<li>
<code>FirstBCTumorSize</code>: Size category of the first primary breast cancer tumor. The only options are <code>"T0/T1/T2"</code>, <code>"T3/T4"</code>, <code>"Tis"</code>. Use <code>NA</code> for unknown tumor size or if there was no first primary breast cancer.</li>
</ul>
</li>
</ul>
<p>To represent a cancer in <code>pedigree</code>, we need to use <code>short</code> cancer codes:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">PanelPRO</span><span class="fu">:::</span><span class="va"><a href="reference/CANCER_NAME_MAP.html">CANCER_NAME_MAP</a></span><span class="op">)</span></span>
<span><span class="co">#&gt;    short                long</span></span>
<span><span class="co">#&gt; 1    BRA               Brain</span></span>
<span><span class="co">#&gt; 2     BC              Breast</span></span>
<span><span class="co">#&gt; 3    COL          Colorectal</span></span>
<span><span class="co">#&gt; 4   ENDO         Endometrial</span></span>
<span><span class="co">#&gt; 5    GAS             Gastric</span></span>
<span><span class="co">#&gt; 6    KID              Kidney</span></span>
<span><span class="co">#&gt; 7   LEUK            Leukemia</span></span>
<span><span class="co">#&gt; 8   MELA            Melanoma</span></span>
<span><span class="co">#&gt; 9     OC             Ovarian</span></span>
<span><span class="co">#&gt; 10   OST        Osteosarcoma</span></span>
<span><span class="co">#&gt; 11  PANC            Pancreas</span></span>
<span><span class="co">#&gt; 12  PROS            Prostate</span></span>
<span><span class="co">#&gt; 13    SI     Small Intestine</span></span>
<span><span class="co">#&gt; 14   STS Soft Tissue Sarcoma</span></span>
<span><span class="co">#&gt; 15   THY             Thyroid</span></span>
<span><span class="co">#&gt; 16    UB     Urinary Bladder</span></span>
<span><span class="co">#&gt; 17   HEP       Hepatobiliary</span></span>
<span><span class="co">#&gt; 18   CBC       Contralateral</span></span></code></pre></div>
<p>For example, <code>BC</code> is the <code>short</code> name that maps to <code>Breast</code>, so breast cancer affection status and age of diagnosis should be recorded as columns named <code>isAffBC</code> and <code>AgeBC</code>.</p>
<p>Note <code>Contralateral</code> indicates contralateral breast cancer (<code>CBC</code>).</p>
<p>Unknown values in <code>pedigree</code> should be explicitly coded as <code>NA</code>.</p>
<p>Before computing anything, the <code>PanelPRO</code> function will check the validity and structural consistency of <code>pedigree</code>. Errors will be corrected when possible, or the function will halt with an informative error message if not.</p>
<p>The <code>PanelPRO</code> package comes with several sample pedigrees, each of which contains family history information. For example, <code>err_fam_1</code> is a pedigree that will fail the validity checks and <code>test_fam_1</code> is a valid pedigree that contains 19 relatives and family history for breast and ovarian cancer:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_fam_1</span><span class="op">)</span></span>
<span><span class="co">#&gt;   ID Sex MotherID FatherID isProband CurAge isAffBC isAffOC AgeBC AgeOC isDead</span></span>
<span><span class="co">#&gt; 1  1   0       NA       NA         0     93       1       0    65    NA      1</span></span>
<span><span class="co">#&gt; 2  2   1       NA       NA         0     80       0       0    NA    NA      1</span></span>
<span><span class="co">#&gt; 3  3   0        1        2         0     72       1       1    40    NA      0</span></span>
<span><span class="co">#&gt; 4  4   1        1        2         0     65       0       0    NA    NA      1</span></span>
<span><span class="co">#&gt; 5  5   1        1        2         0     65       0       0    NA    NA      0</span></span>
<span><span class="co">#&gt; 6  6   0        1        2         1     55       0       0    NA    NA      0</span></span>
<span><span class="co">#&gt;        race      riskmod interAge Twins BRCA1 BRCA2 ER CK5.6 CK14 PR HER2</span></span>
<span><span class="co">#&gt; 1 All_Races                           0    NA    NA NA    NA   NA NA   NA</span></span>
<span><span class="co">#&gt; 2 All_Races                           0     1     0 NA    NA   NA NA   NA</span></span>
<span><span class="co">#&gt; 3 All_Races   Mastectomy       28     0    NA    NA NA    NA   NA NA   NA</span></span>
<span><span class="co">#&gt; 4 All_Races                           1     1    NA NA    NA   NA NA   NA</span></span>
<span><span class="co">#&gt; 5 All_Races                           1    NA    NA NA    NA   NA NA   NA</span></span>
<span><span class="co">#&gt; 6 All_Races Hysterectomy       48     0    NA    NA  0    NA   NA NA   NA</span></span>
<span><span class="co">#&gt;   Ancestry</span></span>
<span><span class="co">#&gt; 1    nonAJ</span></span>
<span><span class="co">#&gt; 2    nonAJ</span></span>
<span><span class="co">#&gt; 3       AJ</span></span>
<span><span class="co">#&gt; 4    nonAJ</span></span>
<span><span class="co">#&gt; 5    nonAJ</span></span>
<span><span class="co">#&gt; 6    nonAJ</span></span></code></pre></div>
<p>For more information on the sample pedigrees, see their individual help pages. For an example of building a pedigree, see the vignette on “Building PanelPRO Pedigrees”.</p>
</div>
<div class="section level3">
<h3 id="model-specification">Model specification<a class="anchor" aria-label="anchor" href="#model-specification"></a>
</h3>
<p>There are a few ways to specify the model to be run by <code>PanelPRO</code>. The first is to pass a character string to <code>model_spec</code> that indicates a pre-specified model. Available options are:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">PanelPRO</span><span class="fu">:::</span><span class="va"><a href="reference/MODELPARAMS.html">MODELPARAMS</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "BRCAPRO"  "BRCAPRO5" "BRCAPRO6" "MMRPRO"   "PanPRO11" "PanPRO22"</span></span></code></pre></div>
<p>The genes and cancers that these models specify can be found by accessing the named list element in <code>PanelPRO:::MODELPARAMS</code>. For example, the 22 genes and 17 cancers specified by <code>"PanPRO22"</code> are:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">PanelPRO</span><span class="fu">:::</span><span class="va"><a href="reference/MODELPARAMS.html">MODELPARAMS</a></span><span class="op">$</span><span class="va">PanPRO22</span></span>
<span><span class="co">#&gt; $GENES</span></span>
<span><span class="co">#&gt;  [1] "ATM"    "BARD1"  "BRCA1"  "BRCA2"  "BRIP1"  "CDH1"   "CDK4"   "CDKN2A"</span></span>
<span><span class="co">#&gt;  [9] "CHEK2"  "EPCAM"  "MLH1"   "MSH2"   "MSH6"   "MUTYH"  "NBN"    "PALB2" </span></span>
<span><span class="co">#&gt; [17] "PMS2"   "PTEN"   "RAD51C" "RAD51D" "STK11"  "TP53"  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $CANCERS</span></span>
<span><span class="co">#&gt;  [1] "Brain"               "Breast"              "Colorectal"         </span></span>
<span><span class="co">#&gt;  [4] "Endometrial"         "Gastric"             "Kidney"             </span></span>
<span><span class="co">#&gt;  [7] "Leukemia"            "Melanoma"            "Ovarian"            </span></span>
<span><span class="co">#&gt; [10] "Osteosarcoma"        "Pancreas"            "Prostate"           </span></span>
<span><span class="co">#&gt; [13] "Small Intestine"     "Soft Tissue Sarcoma" "Thyroid"            </span></span>
<span><span class="co">#&gt; [16] "Urinary Bladder"     "Hepatobiliary"       "Contralateral"</span></span></code></pre></div>
<p>For convenience, the <code>PanelPRO</code> package exports wrapper functions for all of the pre-specified models: <code>BRCAPRO</code>, <code>BRCAPRO5</code>, <code>BRCAPRO6</code>, <code>MMRPRO</code>, <code>PanelPRO11</code>, and <code>PanelPRO22</code>.</p>
<p>Alternatively, the user can explicitly pass in genes to the <code>genes</code> argument and cancers to the <code>cancers</code> argument. If the inputs conflict with each other, <code>genes</code> and <code>cancers</code> will override the <code>model_spec</code> specification. Available genes are:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">PanelPRO</span><span class="fu">:::</span><span class="va"><a href="reference/GENE_TYPES.html">GENE_TYPES</a></span></span>
<span><span class="co">#&gt;  [1] "ATM"    "BARD1"  "BRCA1"  "BRCA2"  "BRIP1"  "CDH1"   "CDK4"   "CDKN2A"</span></span>
<span><span class="co">#&gt;  [9] "CHEK2"  "EPCAM"  "MLH1"   "MSH2"   "MSH6"   "MUTYH"  "NBN"    "PALB2" </span></span>
<span><span class="co">#&gt; [17] "PMS2"   "PTEN"   "RAD51C" "RAD51D" "STK11"  "TP53"</span></span></code></pre></div>
<p>Available cancers are:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">PanelPRO</span><span class="fu">:::</span><span class="va"><a href="reference/CANCER_TYPES.html">CANCER_TYPES</a></span></span>
<span><span class="co">#&gt;  [1] "Brain"               "Breast"              "Colorectal"         </span></span>
<span><span class="co">#&gt;  [4] "Endometrial"         "Gastric"             "Kidney"             </span></span>
<span><span class="co">#&gt;  [7] "Leukemia"            "Melanoma"            "Ovarian"            </span></span>
<span><span class="co">#&gt; [10] "Osteosarcoma"        "Pancreas"            "Prostate"           </span></span>
<span><span class="co">#&gt; [13] "Small Intestine"     "Soft Tissue Sarcoma" "Thyroid"            </span></span>
<span><span class="co">#&gt; [16] "Urinary Bladder"     "Hepatobiliary"       "Contralateral"</span></span></code></pre></div>
<p>Therefore, it is equivalent to run:</p>
<ul>
<li><code>PanelPRO(pedigree, model_spec = "PanPRO22", ...)</code></li>
<li><code>PanelPRO22(pedigree, ...)</code></li>
<li><code>PanelPRO(pedigree, genes = PanelPRO:::MODELPARAMS$PanPRO22$GENES, cancers = PanelPRO:::MODELPARAMS$PanPRO22$CANCERS, ...)</code></li>
</ul>
<p>By default, <code>PanelPRO</code> without any user-specified <code>model_spec</code>, <code>genes</code>, or <code>cancers</code> will run <code>PanelPRO22</code>, so it is also equivalent in this case to simply write <code>PanelPRO(pedigree, ...)</code>. Cancers for which no family history is available in the pedigree will be dropped from the model specification, under the default settings.</p>
</div>
<div class="section level3">
<h3 id="additional-notes">Additional notes<a class="anchor" aria-label="anchor" href="#additional-notes"></a>
</h3>
<ul>
<li><p>The <code>PanelPRO</code> function pulls model parameters, including mutation allele frequencies and cancer penetrances, from the <code>database</code> argument. The internal <code>PanelPRODataBase</code> is used by default, but users can override it by passing in their own database.</p></li>
<li><p>The optional <code>proband</code> argument allows the user to specify the IDs for one or more probands for whom carrier probabilities and future risk should be estimated, overriding the contents of the <code>isProband</code> column in <code>pedigree</code>. If <code>proband = "All"</code>, the results for all individuals in the pedigree will be calculated.</p></li>
<li><p>The <code>PanelPRO</code> function returns a list with two elements, <code>posterior.prob</code> and <code>future.risk</code>. By default, <code>PanelPRO</code> imputes 20 iterations of missing ages (<code>impute.missing.ages = TRUE</code> and <code>iterations = 20</code>). When missing ages are multiply imputed, <code>posterior.prob</code> and <code>future.risk</code> report the average and range of the results from these imputations.</p></li>
<li><p>The <code>max.mut</code> argument sets the maximum number of simultaneous mutations allowed. By default, <code>max.mut</code> will be set to <code>2</code> when <code>pedigree</code> has 30 or fewer members and <code>1</code> when <code>pedigree</code> exceeds 30 members. For example, if <code>max.mut = 2</code>, a model that includes MLH1, MSH2, and MSH6 will estimate the carrier probability of being a carrier of both MLH1 and MSH2 (or of both MLH1 and MSH6, or of both MSH2 and MSH6), but not of being a carrier of MLH1, MSH2, and MSH6 simultaneously. Choosing a large value for <code>max.mut</code>, especially when there is a large number of genes in the model, may result in slow performance.</p></li>
<li><p>The <code>age.by</code> argument determines the fixed age intervals for the future risk calculations. By default, <code>age.by = 5</code>, so if the proband’s current age is 30, future risk will be calculated at ages 35, 40, 45, etc.</p></li>
<li><p>All breast cancers are assumed to be invasive or mixed invasive and DCIS. Pure DCIS breast cancers are not compatible with PanelPRO.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="example-usage">Example usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h3>
<p>All three of the following expressions are equivalent calls that fit the <code>BRCAPRO</code> model to the <code>test_fam_1</code> pedigree. The resulting posterior carrier probabilities and future risk estimates for the proband (taken from the <code>test_fam_1$isProband</code> column to be the individual with ID <code>6</code>) are identical.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Pass in BRCAPRO as the model specification</span></span>
<span><span class="va">out1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PanelPRO.html">PanelPRO</a></span><span class="op">(</span>pedigree <span class="op">=</span> <span class="va">test_fam_1</span>, </span>
<span>                 model_spec <span class="op">=</span> <span class="st">"BRCAPRO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span></span>
<span><span class="co">#&gt; Your model has 2 cancers - Breast, Ovarian and 2 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Setting max.mut to 2.</span></span>
<span><span class="co">#&gt; ID 6,29 has tumor marker testing results but is unaffected for relevant cancer, so testing will be ignored.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA1 are assumed to be for default variant BRCA1_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA2 are assumed to be for default variant BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; ID 3 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 10 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 9 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 16,17 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 29,30 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 6's lower and upper bounds for the probability of carrying any pathogenic variant are (0.30599, 0.30604).</span></span>
<span><span class="co">#&gt; The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span></span>
<span></span>
<span><span class="co"># Explicitly specify the genes and cancers in BRCAPRO</span></span>
<span><span class="va">out2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PanelPRO.html">PanelPRO</a></span><span class="op">(</span>pedigree <span class="op">=</span> <span class="va">test_fam_1</span>, </span>
<span>                 genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"BRCA2"</span><span class="op">)</span>, </span>
<span>                 cancers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Breast"</span>, <span class="st">"Ovarian"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Your model has 2 cancers - Breast, Ovarian and 2 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Setting max.mut to 2.</span></span>
<span><span class="co">#&gt; ID 6,29 has tumor marker testing results but is unaffected for relevant cancer, so testing will be ignored.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA1 are assumed to be for default variant BRCA1_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA2 are assumed to be for default variant BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; ID 3 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 10 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 9 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 16,17 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 29,30 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 6's lower and upper bounds for the probability of carrying any pathogenic variant are (0.30599, 0.30604).</span></span>
<span><span class="co">#&gt; The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">out1</span>, <span class="va">out2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pre-defined model wrapper for BRCAPRO</span></span>
<span><span class="va">out3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/BRCAPRO.html">BRCAPRO</a></span><span class="op">(</span>pedigree <span class="op">=</span> <span class="va">test_fam_1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span></span>
<span><span class="co">#&gt; Your model has 2 cancers - Breast, Ovarian and 2 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Setting max.mut to 2.</span></span>
<span><span class="co">#&gt; ID 6,29 has tumor marker testing results but is unaffected for relevant cancer, so testing will be ignored.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA1 are assumed to be for default variant BRCA1_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Germline testing results for BRCA2 are assumed to be for default variant BRCA2_hetero_anyPV.</span></span>
<span><span class="co">#&gt; ID 3 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 10 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 9 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 16,17 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 29,30 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's race has been changed to All_Races to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 33,34 's Ancestry has been changed to nonAJ to meet hereditary consistency.</span></span>
<span><span class="co">#&gt; ID 6's lower and upper bounds for the probability of carrying any pathogenic variant are (0.30599, 0.30604).</span></span>
<span><span class="co">#&gt; The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">out1</span>, <span class="va">out3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>visRisk</code> function provides interactive visualizations for exploring the proband(s)’s carrier probabilities and future risk estimates returned by <code>PanelPRO</code>. For example,</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/visRisk.html">visRisk</a></span><span class="op">(</span><span class="va">out1</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="reference/figures/ans1.png" alt=""><p class="caption">vis_risk</p>
</div>
<p><code>PanelPRO11</code> runs an 11 gene, 11 cancer model. <code>test_fam_2</code> is a more complex pedigree with 25 relatives and information on endometrial, pancreatic, and small intestine cancer. By default, the cancers that are not included in the pedigree will be dropped from the model specification. Carrier probabilities will therefore be estimated based on family history of endometrial, pancreatic, and small intestine cancer; future risk estimates will only be given for these three cancers.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out4</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PanelPRO11.html">PanelPRO11</a></span><span class="op">(</span>pedigree <span class="op">=</span> <span class="va">test_fam_2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Brain, Breast, Colorectal, Gastric, Kidney, Melanoma, Ovarian, Prostate, Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span></span>
<span><span class="co">#&gt; Your model has 3 cancers - Endometrial, Pancreas, Small Intestine and 11 genes - BRCA1_hetero_anyPV, BRCA2_hetero_anyPV, ATM_hetero_anyPV, PALB2_hetero_anyPV, CHEK2_hetero_1100delC, EPCAM_hetero_anyPV, PMS2_hetero_anyPV, MLH1_hetero_anyPV, MSH2_hetero_anyPV, MSH6_hetero_anyPV, CDKN2A[P16]_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Setting max.mut to 2.</span></span>
<span><span class="co">#&gt; ID 12's lower and upper bounds for the probability of carrying any pathogenic variant are (0.18282, 0.25576).</span></span>
<span><span class="co">#&gt; ID 19's lower and upper bounds for the probability of carrying any pathogenic variant are (0.10647, 0.14387).</span></span>
<span><span class="co">#&gt; ID 21's lower and upper bounds for the probability of carrying any pathogenic variant are (0.04728, 0.04934).</span></span>
<span><span class="co">#&gt; The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span></span></code></pre></div>
<p><code>out2</code> contains results for the three probands specified by the <code>isProband</code> column in <code>fam25</code>. We can extract the results for a specific proband, e.g. the one with <code>ID = 12</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out4</span><span class="op">$</span><span class="va">posterior.prob</span><span class="op">$</span><span class="va">`12`</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   genes     estimate        lower        upper</span></span>
<span><span class="co">#&gt; 1            noncarrier 0.7862760882 0.7442372677 0.8171816929</span></span>
<span><span class="co">#&gt; 2    BRCA1_hetero_anyPV 0.0009420256 0.0009121588 0.0009630147</span></span>
<span><span class="co">#&gt; 3    BRCA2_hetero_anyPV 0.0010629569 0.0010320276 0.0010863401</span></span>
<span><span class="co">#&gt; 4      ATM_hetero_anyPV 0.0030277342 0.0029323624 0.0030951922</span></span>
<span><span class="co">#&gt; 5    PALB2_hetero_anyPV 0.0008971822 0.0008710650 0.0009169285</span></span>
<span><span class="co">#&gt; 6 CHEK2_hetero_1100delC 0.0043207063 0.0041660817 0.0044303642</span></span>
<span></span>
<span><span class="va">out4</span><span class="op">$</span><span class="va">future.risk</span><span class="op">$</span><span class="va">`12`</span></span>
<span><span class="co">#&gt; $Endometrial</span></span>
<span><span class="co">#&gt;    ByAge   estimate      lower      upper</span></span>
<span><span class="co">#&gt; 1     49 0.01133006 0.00953546 0.01382722</span></span>
<span><span class="co">#&gt; 2     54 0.03238989 0.02722494 0.03958295</span></span>
<span><span class="co">#&gt; 3     59 0.06009958 0.05068241 0.07320253</span></span>
<span><span class="co">#&gt; 4     64 0.08249396 0.07018836 0.09955262</span></span>
<span><span class="co">#&gt; 5     69 0.09418568 0.08118970 0.11208450</span></span>
<span><span class="co">#&gt; 6     74 0.10062249 0.08772854 0.11826668</span></span>
<span><span class="co">#&gt; 7     79 0.10498755 0.09222464 0.12237341</span></span>
<span><span class="co">#&gt; 8     84 0.10786272 0.09518549 0.12508302</span></span>
<span><span class="co">#&gt; 9     89 0.10942380 0.09679517 0.12655370</span></span>
<span><span class="co">#&gt; 10    94 0.10999928 0.09739079 0.12709479</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pancreas</span></span>
<span><span class="co">#&gt;    ByAge     estimate        lower        upper</span></span>
<span><span class="co">#&gt; 1     49 0.0005381908 0.0004860642 0.0006117279</span></span>
<span><span class="co">#&gt; 2     54 0.0014148118 0.0012887253 0.0015924263</span></span>
<span><span class="co">#&gt; 3     59 0.0028634411 0.0026154834 0.0032126670</span></span>
<span><span class="co">#&gt; 4     64 0.0051014429 0.0046596368 0.0057239684</span></span>
<span><span class="co">#&gt; 5     69 0.0081994916 0.0074999414 0.0091853703</span></span>
<span><span class="co">#&gt; 6     74 0.0117457437 0.0107899961 0.0130922158</span></span>
<span><span class="co">#&gt; 7     79 0.0153317598 0.0141921358 0.0169357311</span></span>
<span><span class="co">#&gt; 8     84 0.0183264326 0.0170970105 0.0200550145</span></span>
<span><span class="co">#&gt; 9     89 0.0204984458 0.0192487633 0.0222542152</span></span>
<span><span class="co">#&gt; 10    94 0.0215754010 0.0203364406 0.0233155908</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Small Intestine`</span></span>
<span><span class="co">#&gt;    ByAge    estimate       lower       upper</span></span>
<span><span class="co">#&gt; 1     49 0.001426245 0.001165799 0.001792689</span></span>
<span><span class="co">#&gt; 2     54 0.003190522 0.002623869 0.003985722</span></span>
<span><span class="co">#&gt; 3     59 0.005204334 0.004316200 0.006444637</span></span>
<span><span class="co">#&gt; 4     64 0.007393643 0.006176663 0.009086996</span></span>
<span><span class="co">#&gt; 5     69 0.009509772 0.007988545 0.011622333</span></span>
<span><span class="co">#&gt; 6     74 0.011385919 0.009599729 0.013865538</span></span>
<span><span class="co">#&gt; 7     79 0.012799079 0.010814630 0.015556873</span></span>
<span><span class="co">#&gt; 8     84 0.013798228 0.011682139 0.016741988</span></span>
<span><span class="co">#&gt; 9     89 0.014272380 0.012103188 0.017291318</span></span>
<span><span class="co">#&gt; 10    94 0.014381484 0.012207102 0.017407885</span></span></code></pre></div>
<p>Finally, we can run PanelPRO-22 (the default for <code>PanelPRO</code>) on <code>test_fam_3</code>, which has 50 relatives and stores family history for the 11 cancers specified by PanelPRO-11. Note that because there are more than 30 relatives in <code>test_fam_3</code>, the <code>max.mut</code> argument, which sets the maximum number of simultaneous mutations allowed, defaults to <code>1</code> instead of <code>2</code>. We set <code>proband = 10</code>, overriding the proband information encoded in <code>test_fam_3$isProband</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out5</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PanelPRO.html">PanelPRO</a></span><span class="op">(</span>pedigree <span class="op">=</span> <span class="va">test_fam_3</span>, proband <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Leukemia, Osteosarcoma, Soft Tissue Sarcoma, Thyroid, Urinary Bladder, Hepatobiliary, Contralateral cancer(s) not in pedigree, so they will be removed from the model specification.</span></span>
<span><span class="co">#&gt; Your model has 11 cancers - Brain, Breast, Colorectal, Endometrial, Gastric, Kidney, Melanoma, Ovarian, Pancreas, Prostate, Small Intestine and 22 genes - ATM_hetero_anyPV, BARD1_hetero_anyPV, BRCA1_hetero_anyPV, BRCA2_hetero_anyPV, BRIP1_hetero_anyPV, CDH1_hetero_anyPV, CDK4_hetero_anyPV, CDKN2A[P16]_hetero_anyPV, CHEK2_hetero_1100delC, EPCAM_hetero_anyPV, MLH1_hetero_anyPV, MSH2_hetero_anyPV, MSH6_hetero_anyPV, MUTYH_hetero_anyPV, MUTYH_homo_anyPV, NBN_hetero_657del5, PALB2_hetero_anyPV, PMS2_hetero_anyPV, PTEN_hetero_anyPV, RAD51C_hetero_anyPV, RAD51D_hetero_anyPV, STK11_hetero_anyPV, TP53_hetero_anyPV.</span></span>
<span><span class="co">#&gt; Setting max.mut to 1.</span></span>
<span><span class="co">#&gt; `proband` argument does not match with "isProband" column in the pedigree, using `proband`</span></span>
<span><span class="co">#&gt; Riskmods that are NA will be removed</span></span>
<span><span class="co">#&gt; Warning: ID 14 is male but has been affected by OC; turning into non-affection.</span></span>
<span><span class="co">#&gt; ID 10's lower and upper bounds for the probability of carrying any pathogenic variant are (0.9246, 0.97597).</span></span>
<span><span class="co">#&gt; The value ranges in the results are driven by the different imputed missing ages at each imputation, so please consider including more age information when possible to reduce the range widths.</span></span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing PanelPRO</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>BayesMendel Lab <br><small class="roles"> Author, maintainer </small>  </li>
<li>Theo Huang <br><small class="roles"> Author </small>  </li>
<li>Gavin Lee <br><small class="roles"> Author </small>  </li>
<li>Jane Liang <br><small class="roles"> Author </small>  </li>
<li>Tom Madsen <br><small class="roles"> Author </small>  </li>
<li>Qing Zhang <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by BayesMendel Lab, Theo Huang, Gavin Lee, Jane Liang, Tom Madsen, Qing Zhang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
